golang/cmd/kafka-to-postgresql/kafkaSharedTypes.go

```diff
@@ -0,0 +1,184 @@
+			zap.S().Debugf("%s Waiting for put back channel to empty: %d", identifier, len(putBackChannel))
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		var msg *kafka.Message
+		// Wait for new messages
+		// This has a timeout, allowing ShuttingDown to be checked
+		msg, err = kafkaConsumer.ReadMessage(5000)
+		if err != nil {
+			if err.(kafka.Error).Code() == kafka.ErrTimedOut {
+				continue
+			} else if err.(kafka.Error).Code() == kafka.ErrUnknownTopicOrPart {
+				// This will occur when no topic for the regex is available !
+				zap.S().Errorf("%s Unknown topic or partition: %s", identifier, err)
+				ShutdownApplicationGraceful()
+				return
+			} else {
+				zap.S().Warnf("%s Failed to read kafka message: %s: %s", identifier, err, err.(kafka.Error).Code())
+				ShutdownApplicationGraceful()
+				return
+			}
+		}
+		// Insert received message into the processor channel
+		processorChannel <- msg
+		// Increase message counter, for stats
+		Messages += 1
+	}
+	zap.S().Debugf("%s Shutting down Kafka consumer for topic %s", identifier, topic)
+}
+
+// startPutbackProcessor starts the putback processor.
+// It will put unprocessable messages back into the kafka queue, modifying there key to include the reason and error.
+func startPutbackProcessor(identifier string, putBackChannel chan PutBackChanMsg, kafkaProducer *kafka.Producer) {
+	// Loops until the shutdown signal is received and the channel is empty
+	for !ShutdownPutback {
+		select {
+		case msgX := <-putBackChannel:
+			{
+				current := time.Now().UnixMilli()
+				var msg = msgX.msg
+				var reason = msgX.reason
+				var errorString = msgX.errorString
+
+				if msg == nil {
+					continue
+				}
+
+				var kafkaKey KafkaKey
+				if msg.Key == nil {
+					kafkaKey = KafkaKey{
+						&Putback{
+							FirstTsMS: current,
+							LastTsMS:  current,
+							Amount:    1,
+							Reason:    reason,
+						},
+					}
+				} else {
+					err := jsoniter.Unmarshal(msg.Key, &kafkaKey)
+					if err != nil {
+						kafkaKey = KafkaKey{
+							&Putback{
+								FirstTsMS: current,
+								LastTsMS:  current,
+								Amount:    1,
+								Reason:    reason,
+							},
+						}
+					} else {
+						kafkaKey.Putback.LastTsMS = current
+						kafkaKey.Putback.Amount += 1
+						kafkaKey.Putback.Reason = reason
+					}
+				}
+
+				if errorString != nil && *errorString != "" {
+					kafkaKey.Putback.Error = *errorString
+				}
+
+				var err error
+				msg.Key, err = jsoniter.Marshal(kafkaKey)
+				if err != nil {
+					zap.S().Errorf("%s Failed to marshal key: %v (%s)", identifier, kafkaKey, err)
+					err = nil
+				}
+
+				msgx := kafka.Message{
+					TopicPartition: kafka.TopicPartition{
+						Topic:     msg.TopicPartition.Topic,
+						Partition: msg.TopicPartition.Partition,
+					},
+					Value: msg.Value,
+					Key:   msg.Key,
+				}
+
+				err = kafkaProducer.Produce(&msgx, nil)
```

Comment:
Delivery notifications have to be handled, else kafka will stop after ~1mil messages